omniscidb/QueryBuilder/QueryBuilder.h

```diff
@@ -0,0 +1,656 @@
+  BuilderExpr parseAggString(const std::string& agg_str) const;
+  std::vector<BuilderExpr> parseAggString(const std::vector<std::string>& aggs) const;
+
+  const QueryBuilder* builder_;
+  NodePtr node_;
+};
+
+class QueryBuilder {
+ public:
+  QueryBuilder(Context& ctx, SchemaProviderPtr schema_provider, ConfigPtr config);
+
+  BuilderNode scan(const std::string& table_name) const;
+  BuilderNode scan(int db_id, const std::string& table_name) const;
+  BuilderNode scan(int db_id, int table_id) const;
+  BuilderNode scan(const TableRef& table_ref) const;
+
+  BuilderExpr count() const;
+
+  BuilderExpr cst(int val) const;
+  BuilderExpr cst(int val, const Type* type) const;
+  BuilderExpr cst(int val, const std::string& type) const;
+  BuilderExpr cst(int64_t val) const;
+  BuilderExpr cst(int64_t val, const Type* type) const;
+  BuilderExpr cst(int64_t val, const std::string& type) const;
+  BuilderExpr cst(double val) const;
+  BuilderExpr cst(double val, const Type* type) const;
+  BuilderExpr cst(double val, const std::string& type) const;
+  BuilderExpr cst(const std::string& val) const;
+  BuilderExpr cst(const std::string& val, const Type* type) const;
+  BuilderExpr cst(const std::string& val, const std::string& type) const;
+  BuilderExpr cst(std::initializer_list<int>& vals) const;
+  BuilderExpr cst(std::initializer_list<int> vals, const Type* type) const;
+  BuilderExpr cst(std::initializer_list<int> vals, const std::string& type) const;
+  BuilderExpr cst(std::initializer_list<double> vals) const;
+  BuilderExpr cst(std::initializer_list<double> vals, const Type* type) const;
+  BuilderExpr cst(std::initializer_list<double> vals, const std::string& type) const;
+  BuilderExpr cst(const std::vector<int>& vals) const;
+  BuilderExpr cst(const std::vector<int>& vals, const Type* type) const;
+  BuilderExpr cst(const std::vector<int>& vals, const std::string& type) const;
+  BuilderExpr cst(const std::vector<double>& vals) const;
+  BuilderExpr cst(const std::vector<double>& vals, const Type* type) const;
+  BuilderExpr cst(const std::vector<double>& vals, const std::string& type) const;
+  BuilderExpr cst(const std::vector<std::string>& vals, const Type* type) const;
+  BuilderExpr cst(const std::vector<std::string>& vals, const std::string& type) const;
+  BuilderExpr cstNoScale(int64_t val, const Type* type) const;
+  BuilderExpr cstNoScale(int64_t val, const std::string& type) const;
+  BuilderExpr trueCst() const;
+  BuilderExpr falseCst() const;
+  BuilderExpr nullCst() const;
+  BuilderExpr nullCst(const Type* type) const;
+  BuilderExpr nullCst(const std::string& type) const;
+  BuilderExpr date(const std::string& val) const;
+  BuilderExpr time(const std::string& val) const;
+  BuilderExpr timestamp(const std::string& val) const;
+
+ protected:
+  friend class BuilderExpr;
+  friend class BuilderNode;
+
+  BuilderNode scan(TableInfoPtr table_info) const;
+
+  Context& ctx_;
```

Comment:
Is it possible to reuse a chain of operators built with a different builder? Assuming they share the schema and data source.

Reply:
IR contexts were introduced for the isolation of IRs built in different threads. If you work on the same thread and have two builders working with the same context, then you can mix their expressions, but I don't see why you would need it.

So far, we don't actually manage contexts nicely in our code and always use the same default context. That means we shouldn't try to build multiple queries from different threads. I hope it will change some days and we start creating query-scoped contexts. For Python, I think we will continue using a single global context for all queries.